File: CSharp/DomWalker/DomDocument.cs
using System.IO;
using System.Text;
using static HtmlWalker.Platform.WalkerPlatform;

namespace DomWalker
{
    public abstract class DomDocument<F, A, D> : WalkerDocument where F : DomFactory<A, D> where A : DomApi<D>
    {
        private readonly F factory;

        protected DomDocument(F factory) : base() { this.factory = factory; }

        protected D Dom { get => this.factory.Dom; set => this.factory.Dom = value; }

        protected override object DocumentNode
            => this.factory.DApi.DocumentNode(Dom);

        public override string Html => this.factory.DApi.Html(Dom);

        public override WalkerFactory Factory => factory;

        public override WalkerDocument Load(string path)
        {
            this.factory.DApi.Load(Dom, path);
            return this;
        }

        public override WalkerDocument Load(TextReader reader)
        {
            this.factory.DApi.Load(Dom, reader);
            return this;
        }

        public override WalkerDocument LoadHtml(string html)
        {
            this.factory.DApi.LoadHtml(Dom, html);
            return this;
        }

        public override void Save(string path, Encoding encoding) => this.factory.DApi.Save(Dom, path, Encoding);

        public override void Save(string path) => this.factory.DApi.Save(Dom, path);

        public override void Save(TextWriter writer) => this.factory.DApi.Save(Dom, writer);

        public override void Close()
        {
            base.Close();
            this.factory.Close();
        }
    }
}

File: Java/htmlwalker/src/main/java/domwalker/DomDocument.java
package domwalker;

import java.io.Writer;
import java.nio.charset.Charset;

import htmlwalker.exception.HtmlWalkerException;
import htmlwalker.platform.WalkerDocument;
import htmlwalker.platform.WalkerFactory;

public abstract class DomDocument<F extends DomFactory<A, D>, A extends DomApi<D>, D> extends WalkerDocument
{
	private final F factory;
		
	protected DomDocument(F factory)
	{
		super();
		this.factory = factory;
	}
	
	protected D getDom() { return this.factory.dom; }
	protected void setDom(D dom) { this.factory.dom = dom; }
	
	@Override
	protected Object documentNode() { return this.factory.dapi.documentNode(getDom()); }

	@Override
	public WalkerFactory factory() { return this.factory; }
	
	@Override
	public String html() { return this.factory.dapi.html(getDom()); }

	@Override
    public WalkerDocument load(String path) throws HtmlWalkerException
    {
		close();
		
        this.factory.dom = this.factory.dapi.load(getDom(), path);
        this.documentTag = null;
        return this;
    }


	@Override
    public WalkerDocument loadHtml(String html) throws HtmlWalkerException
    {
		close();

        this.factory.dom = this.factory.dapi.loadHtml(getDom(), html);
        this.documentTag = null;
        return this;
    }
	
	@Override
	public void save(String path, String encoding) throws HtmlWalkerException
	{
		this.factory.dapi.save(getDom(), path, encoding);
	}
	
	@Override
	public void save(String path) throws HtmlWalkerException
	{
		if (this.encoding == null)
			this.factory.dapi.save(getDom(), path);
		else
			this.factory.dapi.save(getDom(), path, this.encoding);
	}
	
	@Override
	public void save(Writer writer) throws HtmlWalkerException
	{
		this.factory.dapi.save(getDom(), writer);
	}
	
	@Override
	public void close()
	{
		super.close();
		this.factory.close();
	}
}
